### PR TITLE
vote offchain data support

### PR DESCRIPTION
**Pull request by kderme**
# Description

Partial implementation of CIp-100 and CIP-108

# Checklist

- [x] Commit sequence broadly makes sense
- [ ] Commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] Any changes are noted in the [changelog](https://github.com/IntersectMBO/cardano-db-sync/blob/master/cardano-db-sync/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) on version 0.10.1.0 (which can be run with `scripts/fourmolize.sh`)
- [ ] Self-reviewed the diff

# Migrations

- [x] The pr causes a [breaking change](https://github.com/IntersectMBO/cardano-db-sync/blob/master/doc/migrations.md) of type a,b or c
- [x] If there is a breaking change, the pr includes a database migration and/or a fix process for old values, so that upgrade is possible
- [ ] Resyncing and running the migrations provided will result in the same database semantically

b type migration to add offchain vote data support

### CODE REVIEW

File: cardano-db-sync/src/Cardano/DbSync/Types.hs [hunk 1]
```diff
@@ -141,14 +144,13 @@ data OffChainPoolResult
   | OffChainPoolResultError !OffChainPoolFetchError
 
 data OffChainVoteResult
-  = OffChainVoteResultMetadata !OffChainVoteData
+  = OffChainVoteResultMetadata
+      !OffChainVoteData
+      (DB.OffChainVoteDataId -> [OffChainVoteAuthor])
+      (DB.OffChainVoteDataId -> [OffChainVoteReference])
+      (DB.OffChainVoteDataId -> [OffChainVoteExternalUpdate])
```
Comment (Cmdv) [inline cardano-db-sync/src/Cardano/DbSync/Types.hs:151]: To help with naming and readability could this not be something like:
```
data OffChainVoteResult
  = OffChainVoteResultMetadata !OffChainVoteData OffChainVoteAccessors
  | OffChainVoteResultError !OffChainVoteFetchError
 
 dataOffChainVoteAccessors = OffChainVoteAccessors
  { offChainVoteAuthors :: DB.OffChainVoteDataId -> [OffChainVoteAuthor]
  , offChainVoteReferences :: DB.OffChainVoteDataId -> [OffChainVoteReference]
  , offChainVoteExternalUpdates :: DB.OffChainVoteDataId -> [OffChainVoteExternalUpdate]
  }
  ```
  The call site would look something like:
  ```
  insert = \case
  OffChainVoteResultMetadata md accessors -> do
    mocvdId <- DB.insertOffChainVoteData md
    whenJust mocvdId $ \ocvdId -> do
      DB.insertOffChainVoteAuthors $ offChainVoteAuthors accessors ocvdId
      DB.insertOffChainVoteReference $ offChainVoteReferences accessors ocvdId
      DB.insertOffChainVoteExternalUpdate $ offChainVoteExternalUpdates accessors ocvdId
   ```

File: cardano-db-sync/src/Cardano/DbSync/Types.hs [hunk 2]
```diff
@@ -198,16 +198,14 @@ data OffChainUrlType
...
-
 instance Show OffChainUrlType where
-  show =
-    \case
-      OffChainPoolUrl url -> show url
-      OffChainVoteUrl url -> show url
+  show = getUrl
+
+getUrl :: OffChainUrlType -> String
+getUrl =
```
Comment (Cmdv) [inline cardano-db-sync/src/Cardano/DbSync/Types.hs:205]: could this be called `stringifyUrl` or `showUrl` just that the word "get" doesn't reflect that it's to do with converting a type to a string.

**Review by Cmdv — CHANGES_REQUESTED** [review]

**Comment by kderme** [comment]
@Cmdv addressed and fixed a bug with `getUrl`.

**Review by Cmdv — DISMISSED** [review]
Great thanks @kderme looks 👌 to me

File: cardano-db-sync/src/Cardano/DbSync/OffChain/Vote/Types.hs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+getHashAlgorithm = \case
+  OffChainVoteDataOther dt -> textValue $ hashAlgorithm dt
+  OffChainVoteDataGa dt -> textValue $ hashAlgorithm dt
+
+getLanguage :: OffChainVoteData -> Text
+getLanguage = \case
+  OffChainVoteDataOther dt -> language $ context dt
+  OffChainVoteDataGa dt -> language $ context dt
+
+data OffChainVoteDataTp tp = OffChainVoteDataTp
```
Comment (sgillespie) [inline cardano-db-sync/src/Cardano/DbSync/OffChain/Vote/Types.hs:71]: What's TP? Can you add a comment that tells us?

File: cardano-db-sync/src/Cardano/DbSync/OffChain/Vote/Types.hs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+  deriving (Show, Generic, FromJSON)
+
+data MinimalBody = Body
+  { references :: Maybe [Reference]
+  , comment :: Maybe TextValue
+  , externalUpdates :: Maybe [ExternalUpdate]
+  }
+  deriving (Show, Generic, FromJSON)
+
+data GABody = GABody
```
Comment (sgillespie) [inline cardano-db-sync/src/Cardano/DbSync/OffChain/Vote/Types.hs:106]: I presume this is also _Governance Action_ Body?

**Pull request merged**